### DISCUSSION TITLE
kubernetes configuration for multiple shards CH cluster with a single keeper cluster

### DISCUSSION THREAD

**Discussion by f3r73ch** [ClickHouse/ClickHouse#57473] [General]
Hello

We are having some trouble deploying a clickhouse db in Kubernetes with the following configuration :

- clickhouse cluster with 2 shards and 3 replicas
- one single ch keeper cluster (comprised of 6 nodes). We are running ch keeper in its embedded form.

We were seeking to provide the ability to run ON CLUSTER ddl, across all nodes of the ch db cluster.

For the sake of this example, the following query was used as acid test = "create database test_ch ON cluster default;"

##

The ensemble of k8s manifests used in this example is attached. There are 2 file sets:

1/ one with a ch db with 2 shards and 3 replicas, and 2 keeper clusters (each for each shard) with 3 nodes each. These are the outcomes of exporting the k8s manifest from Bitnami's helm chart.

2/The second defines the same ch db cluster above but with a SINGLE **keeper** cluster, with 6 nodes (2 shards x 3 replicas)

<hr/>
###

1/ For the 1st setup, i.e a ch db with 2 shards x  3 replicas and 2 keeper clusters, all ddl seems to remain constrained to each shard. This means that running the "create database ON CLUSTER ;" creates a database **only** in the shard of the node, where the query is executed, belongs to. So if the query executes in shard 0-0 (1st shard, 1st replica), the query executes immediately for shards 0-0, 0-1, and 0-2 but times out for the nodes belonging to the other shard. (1-0, 1-1 and 1-2)

**Now**, if we amend the sts definition for each shard (in the just created cluster)

 and populate the envs vars pointing to the KEEPER nodes with all the keeper nodes, i.e. , replacing :

- name: KEEPER_NODE_0
value: test-ch-clickhouse-shard0-0.test-ch-clickhouse-headless.f3r304.svc.cluster.local
- name: KEEPER_NODE_1
value: test-ch-clickhouse-shard0-1.test-ch-clickhouse-headless.f3r304.svc.cluster.local
- name: KEEPER_NODE_2
value: test-ch-clickhouse-shard0-2.test-ch-clickhouse-headless.f3r304.svc.cluster.local

with :

- name: KEEPER_NODE_0
value: test-ch-clickhouse-shard0-0.test-ch-clickhouse-headless.f3r304.svc.cluster.local
- name: KEEPER_NODE_1
value: test-ch-clickhouse-shard0-1.test-ch-clickhouse-headless.f3r304.svc.cluster.local
- name: KEEPER_NODE_2
value: test-ch-clickhouse-shard0-2.test-ch-clickhouse-headless.f3r304.svc.cluster.local
- name: KEEPER_NODE_3
value: test-ch-clickhouse-shard1-0.test-ch-clickhouse-headless.f3r304.svc.cluster.local
- name: KEEPER_NODE_4
value: test-ch-clickhouse-shard1-1.test-ch-clickhouse-headless.f3r304.svc.cluster.local
- name: KEEPER_NODE_5
value: test-ch-clickhouse-shard1-2.test-ch-clickhouse-headless.f3r304.svc.cluster.local

then we can create the database using ON CLUSTER and it is visible from all nodes in the CH cluster.

2/ But, if we try to do the last, but from scratch, i.e. starting with a keeper cluster of all 6 nodes, by,, changing the env vars pointing to the KEEPER cluster, the CH cluster never forms correctly (pods keep crashing)

I'm guessing, here, that the initialization of CH requires consistency for the configuration, which we are missing somewhere.

My question is what is the right configuration to have running in k8s a CH db cluster with just a single Keeper cluster that allows for cross-shards DDL ? 

Thank you for your help.

BR, FG 

[ch.issues.zip](https://github.com/ClickHouse/ClickHouse/files/13545528/ch.issues.zip)

**Comment by den-crane** [comment]
we don't support Bitnami's helm chart here
